### PR TITLE
improve SslStream tests

### LINKED ISSUE #1732: Hang running System.Net.Security.Tests.SslStreamTests.DefaultHandshakePipeAsync
I got a hang on one of my Windows machines about 10 hours in. I have dump, logs, etc in `\\danmose3\public\foradam\hang`	

It's waiting for this benchmark to complete
`[2021/03/16 20:07:39][INFO] // Execute: dotnet "c8f219de-c08c-44da-8d8e-cb6168618ed0.dll" --benchmarkName "System.Net.Security.Tests.SslStreamTests.DefaultHandshakePipeAsync" --job "PowerPlanMode=00000000-0000-0000-0000-000000000000, Runtime=.NET 5.0, Arguments=/p:DebugType=portable, Toolchain=net5.0, IterationTime=250.0000 ms, MaxIterationCount=20, MinIterationCount=15, WarmupCount=1" --benchmarkId 2100 in C:\git\performance\artifacts\bin\MicroBenchmarks\Release\net5.0\c8f219de-c08c-44da-8d8e-cb6168618ed0\bin\Release\net5.0`

https://github.com/dotnet/performance/blob/e285e027fe59826d8e60eda270f4615c6baf93c1/src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.cs#L108-L126

just one managed thread, which I guess is waiting on the async test method
```
000000FE78F7D7F8 00007ff90ede29d4 [HelperMethodFrame_1OBJ: 000000fe78f7d7f8] System.Threading.Monitor.ObjWait(Boolean, Int32, System.Object)
000000FE78F7D920 00007ff8d67b6cd3 System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)
000000FE78F7D9C0 00007ff8d67ca939 System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)
000000FE78F7DA30 00007ff8d67ca7e7 System.Threading.Tasks.Task.InternalWaitCore(Int32, System.Threading.CancellationToken)
000000FE78F7DAB0 00007ff8d683c484 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
000000FE78F7DAE0 00007ff8d683c42c System.Runtime.CompilerServices.TaskAwaiter.GetResult()
000000FE78F7DB10 00007ff87763bc24 BenchmarkDotNet.Autogenerated.Runnable_2100.<.ctor>b__3_2()
000000FE78F7DB50 00007ff87763bbc2 BenchmarkDotNet.Autogenerated.Runnable_2100.WorkloadActionNoUnroll(Int64)
000000FE78F7DB90 00007ff87763a07e BenchmarkDotNet.Engines.Engine.RunIteration(BenchmarkDotNet.Engines.IterationData)
000000FE78F7DCD0 00007ff877639acc BenchmarkDotNet.Engines.EngineFactory.Jit(BenchmarkDotNet.Engines.Engine, Int32, Int32, Int32)
```
!dumpasync gives me
```
         Address               MT     Size      State Description
00000210857b5c58 00007ff877722040      208          3 System.Net.Security.SslStream+<ForceAuthenticationAsync>d__171`1[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security]]
00000210857b5dd0 00007ff87776a148      184          0 System.Net.Security.Tests.SslStreamTests+<DefaultHandshake>d__24
```

I can then dump those objects, but I'm really not sure how to proceed. @wfurt can you figure anything more out from this dump? This is your test and I think you know this code best.

cc @adamsitnik

### PR DESCRIPTION
**Pull request by wfurt**
- fixes #1732 - removed dependency on NamePipes and use MemoryStream based approach (copied from runtime) 
- Add test for certificate context to 5.0+ build (shows negligible impact in normal case)
- rework handshake tests. Tls13 will light up on platform where supported. Seems mostly on-par with Tls12
- added basic test for X509Chain.Build. @bartonjs should be happy that Windows & Linux as somewhat close

The default Contonso cert is 4K RSA. The difference is that we use 'localhost' instead of Guid.NewGuid() to allow TLS resume when possible.

WriteReadAsync & ReadWriteAsync looks similar on Linux but shows large difference on Windows.
ConcurrentReadWrite shows big win on Windows.

Somewhat comparable machines
Linux

|                           Method | protocol |        Mean |      Error |     StdDev |      Median |          Min |         Max |   Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |--------- |------------:|-----------:|-----------:|------------:|-------------:|------------:|--------:|------:|------:|----------:|
| DefaultHandshakeContextIPv4Async |        ? | 7,156.14 us |  48.273 us |  45.155 us | 7,152.65 us | 7,062.761 us | 7,237.32 us |       - |     - |     - |  19,495 B |
| DefaultHandshakeContextIPv6Async |        ? | 7,139.02 us |  91.426 us |  85.520 us | 7,155.97 us | 6,962.147 us | 7,272.00 us |       - |     - |     - |  19,495 B |
|        DefaultHandshakeIPv4Async |        ? | 7,490.63 us |  49.744 us |  46.531 us | 7,493.67 us | 7,405.312 us | 7,562.99 us |       - |     - |     - |  21,927 B |
|        DefaultHandshakeIPv6Async |        ? | 7,512.70 us |  71.701 us |  67.069 us | 7,533.24 us | 7,311.420 us | 7,584.27 us |       - |     - |     - |  21,927 B |
|        DefaultHandshakePipeAsync |        ? | 7,499.33 us |  25.540 us |  22.640 us | 7,494.00 us | 7,462.251 us | 7,544.42 us |       - |     - |     - |  22,254 B |
|                   WriteReadAsync |        ? |    12.96 us |   1.989 us |   2.291 us |    13.04 us |     9.421 us |    15.66 us |       - |     - |     - |         - |
|                   ReadWriteAsync |        ? |    16.45 us |   0.323 us |   0.302 us |    16.35 us |    16.043 us |    17.04 us |  0.0400 |     - |     - |     200 B |
|              ConcurrentReadWrite |        ? |    17.24 us |   0.266 us |   0.235 us |    17.25 us |    16.825 us |    17.64 us |       - |     - |     - |       1 B |
|   ConcurrentReadWriteLargeBuffer |        ? |    25.39 us |   0.503 us |   0.538 us |    25.33 us |    24.411 us |    26.36 us |       - |     - |     - |       1 B |
|           HandshakeContonsoAsync |    Tls12 | 7,148.66 us |  25.203 us |  23.575 us | 7,149.89 us | 7,114.260 us | 7,196.38 us |       - |     - |     - |  23,066 B |
 |       HandshakeECDSA256CertAsync |    Tls12 | 2,114.24 us |  19.769 us |  18.492 us | 2,114.55 us | 2,084.222 us | 2,142.41 us |       - |     - |     - |  19,837 B |
 |       HandshakeECDSA512CertAsync |    Tls12 | 4,712.01 us |  78.155 us |  73.107 us | 4,745.00 us | 4,495.792 us | 4,784.19 us |       - |     - |     - |  20,068 B |
|        HandshakeRSA1024CertAsync |    Tls12 | 2,039.34 us | 120.790 us | 139.102 us | 2,080.71 us | 1,499.354 us | 2,106.39 us |       - |     - |     - |  20,029 B |
|        HandshakeRSA2048CertAsync |    Tls12 | 2,384.05 us |  47.187 us |  52.448 us | 2,404.89 us | 2,251.624 us | 2,446.44 us |       - |     - |     - |  20,450 B |
 |        HandshakeRSA4096CertAsync |    Tls12 | 7,140.19 us |  41.522 us |  38.840 us | 7,118.65 us | 7,094.895 us | 7,202.34 us |       - |     - |     - |  21,257 B |
|           HandshakeContonsoAsync |    Tls13 | 7,150.14 us |  26.882 us |  23.830 us | 7,145.94 us | 7,122.749 us | 7,196.83 us |       - |     - |     - |  23,150 B |
|       HandshakeECDSA256CertAsync |    Tls13 | 2,387.83 us | 163.342 us | 188.105 us | 2,447.07 us | 1,719.033 us | 2,503.07 us |       - |     - |     - |  20,000 B |
|       HandshakeECDSA512CertAsync |    Tls13 | 4,751.87 us | 511.595 us | 589.154 us | 5,027.33 us | 3,586.057 us | 5,131.15 us |       - |     - |     - |  20,235 B |
|        HandshakeRSA1024CertAsync |    Tls13 | 1,948.60 us |  31.409 us |  27.844 us | 1,956.73 us | 1,853.786 us | 1,964.80 us |       - |     - |     - |  20,168 B |
|        HandshakeRSA2048CertAsync |    Tls13 | 2,274.49 us |  65.594 us |  75.538 us | 2,299.75 us | 2,051.594 us | 2,318.10 us |       - |     - |     - |  20,579 B |
|        HandshakeRSA4096CertAsync |    Tls13 | 7,296.28 us |  68.566 us |  64.137 us | 7,264.38 us | 7,175.561 us | 7,375.79 us |       - |     - |     - |  21,431 B |

Windows (Insider Preview)
|                           Method | protocol |         Mean |       Error |      StdDev |       Median |          Min |          Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |--------- |-------------:|------------:|------------:|-------------:|-------------:|-------------:|-------:|------:|------:|----------:|
| DefaultHandshakeContextIPv4Async |        ? | 2,082.133 us |  63.4384 us |  73.0557 us | 2,067.205 us | 1,977.345 us | 2,222.273 us |      - |     - |     - |  13,842 B |
| DefaultHandshakeContextIPv6Async |        ? | 2,094.334 us |  38.6982 us |  36.1983 us | 2,104.468 us | 2,022.059 us | 2,136.124 us |      - |     - |     - |  13,833 B |
|        DefaultHandshakeIPv4Async |        ? | 2,499.827 us |  62.6987 us |  72.2039 us | 2,498.695 us | 2,376.656 us | 2,630.068 us |      - |     - |     - |  14,965 B |
|        DefaultHandshakeIPv6Async |        ? | 2,476.645 us |  48.8660 us |  56.2742 us | 2,472.129 us | 2,390.123 us | 2,585.776 us |      - |     - |     - |  14,943 B |
|                   WriteReadAsync |        ? |     6.559 us |   0.1574 us |   0.1813 us |     6.583 us |     6.268 us |     6.969 us |      - |     - |     - |         - |
|                   ReadWriteAsync |        ? |    28.698 us |   0.5409 us |   0.5554 us |    28.600 us |    27.832 us |    29.703 us | 0.0200 |     - |     - |     200 B |
|              ConcurrentReadWrite |        ? |     8.197 us |   0.2750 us |   0.3057 us |     8.138 us |     7.707 us |     8.733 us |      - |     - |     - |       1 B |
|   ConcurrentReadWriteLargeBuffer |        ? |    11.025 us |   0.2180 us |   0.2040 us |    11.075 us |    10.591 us |    11.391 us |      - |     - |     - |         - |
|           HandshakeContonsoAsync |    Tls12 | 7,468.113 us | 198.8782 us | 229.0285 us | 7,454.240 us | 7,142.984 us | 7,960.365 us |      - |     - |     - |  18,555 B |
|       HandshakeECDSA256CertAsync |    Tls12 | 4,427.496 us | 146.7722 us | 169.0230 us | 4,437.419 us | 4,129.809 us | 4,713.367 us |      - |     - |     - |  17,166 B |
|        HandshakeRSA1024CertAsync |    Tls12 | 3,401.830 us |  91.7910 us | 105.7066 us | 3,404.592 us | 3,199.938 us | 3,559.162 us |      - |     - |     - |  17,357 B |
|        HandshakeRSA2048CertAsync |    Tls12 | 3,880.833 us | 131.4012 us | 151.3218 us | 3,921.123 us | 3,681.640 us | 4,091.065 us |      - |     - |     - |  17,741 B |
|        HandshakeRSA4096CertAsync |    Tls12 | 7,232.160 us | 252.7567 us | 291.0750 us | 7,287.827 us | 6,853.891 us | 7,655.282 us |      - |     - |     - |  18,526 B |
|           HandshakeContonsoAsync |    Tls13 | 6,571.124 us | 285.1690 us | 328.4011 us | 6,553.434 us | 6,007.276 us | 7,199.442 us |      - |     - |     - |  25,161 B |
|       HandshakeECDSA256CertAsync |    Tls13 | 3,275.172 us |  89.8350 us | 103.4541 us | 3,290.068 us | 3,078.996 us | 3,419.361 us |      - |     - |     - |  19,355 B |
|        HandshakeRSA1024CertAsync |    Tls13 | 2,470.299 us |  88.0410 us | 101.3881 us | 2,478.413 us | 2,320.644 us | 2,624.479 us |      - |     - |     - |  20,049 B |
|        HandshakeRSA2048CertAsync |    Tls13 | 3,114.217 us | 210.7631 us | 242.7151 us | 3,070.267 us | 2,680.474 us | 3,720.778 us |      - |     - |     - |  21,340 B |
|        HandshakeRSA4096CertAsync |    Tls13 | 6,442.251 us | 376.3603 us | 433.4171 us | 6,400.799 us | 5,907.437 us | 7,350.480 us |      - |     - |     - |  24,162 B |

Linux
|                   Method |     Mean |   Error |  StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------- |---------:|--------:|--------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| BuildX509ChainSelfSigned | 343.4 us | 1.26 us | 1.18 us | 343.8 us | 341.8 us | 345.1 us |     - |     - |     - |      2 KB |
|   BuildX509ChainContonso | 253.2 us | 2.60 us | 2.31 us | 252.9 us | 248.5 us | 257.9 us |     - |     - |     - |      2 KB |

Windows
|                   Method |     Mean |   Error |   StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------- |---------:|--------:|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| BuildX509ChainSelfSigned | 288.7 us | 7.22 us |  8.03 us | 288.1 us | 273.6 us | 304.6 us |     - |     - |     - |     705 B |
|   BuildX509ChainContonso | 353.6 us | 9.39 us | 10.81 us | 350.4 us | 338.2 us | 371.5 us |     - |     - |     - |     717 B |

cc: @stephentoub @geoffkizer @alnikola @adamsitnik @davidfowl

### CODE REVIEW

File: src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Protocols.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            yield return new object[] { SslProtocols.Tls12 };
+            if (SupportsTls13)
+            {
+                yield return new object[] { SslProtocols.Tls13 };
+            }
+        }
+
+        [Benchmark]
+        [ArgumentsSource(nameof(TlsProtocols))]
+        public Task HandshakeContonsoAsync(SslProtocols protocol) => SslStreamTests.HandshakeAsync(SslStreamTests._cert, protocol);
```
Comment (danmoseley) [inline src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Protocols.cs:26]:
```suggestion
        public Task HandshakeContosoAsync(SslProtocols protocol) => SslStreamTests.HandshakeAsync(SslStreamTests._cert, protocol);
```

File: src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Protocols.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+using System.Collections.Generic;
```
Comment (danmoseley) [inline src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Protocols.cs:1]: license header

Comment (wfurt) [reply src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Protocols.cs:5]: updated.

File: src/benchmarks/micro/libraries/System.Security.Cryptography.X509Certificates/X509ChainTests.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        {
+            X509Chain chain = new X509Chain();
+            return chain.Build(certificate);
+        }
+
+        [Benchmark]
+        public bool BuildX509ChainSelfSigned() => BuildChain(_rsa4096Cert);
+
+        [Benchmark]
+        public bool BuildX509ChainContonso() => BuildChain(_cert);
```
Comment (danmoseley) [inline src/benchmarks/micro/libraries/System.Security.Cryptography.X509Certificates/X509ChainTests.cs]:
```suggestion
        public bool BuildX509ChainContoso() => BuildChain(_cert);
```

File: src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Context.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+            using (var sslClient = new SslStream(client, leaveInnerStreamOpen: true, delegate { return true; }))
+            using (var sslServer = new SslStream(server, leaveInnerStreamOpen: true, delegate { return true; }))
+            {
+                await Task.WhenAll(
+                    sslClient.AuthenticateAsClientAsync("localhost", null, SslProtocols.None, checkCertificateRevocation: false),
+                    sslServer.AuthenticateAsServerAsync(serverOptions, default));
+
+                // In Tls1.3 part of handshake happens with data exchange.
+                // To be consistent we do this extra step for all protocol versions
```
Comment (danmoseley) [inline src/benchmarks/micro/libraries/System.Net.Security/SslStreamTests.Context.cs:53]: In SslStreamTests.cs below, it only does this part for TLS1.3.

**Comment by DrewScoggins** [comment]
Just to make sure I understand this PR correctly. We made some changes to how the test works, but it looks like all the tests that changed have gotten new names. Is this correct? If it is then as long as you are satisfied @wfurt these LGTM.

**Comment by wfurt** [comment]
yah. they changed little bit as we have more parametrization. I did sync and resolve conflict with the change @adamsitnik put in.

**Review by adamsitnik — APPROVED** [review]
LGMT, big thanks for improving the benchmarks @wfurt !

**Pull request merged**